src/main/kotlin/eu/europa/ec/eudi/sdjwt/vc/SdJwtVcVerifier.kt

```diff
@@ -0,0 +1,229 @@
+import java.security.cert.X509Certificate
+
+fun interface X509CertificateTrust {
+    suspend fun isTrusted(chain: List<X509Certificate>): Boolean
+}
+
+fun interface DIDResolver {
+    suspend fun resolve(didUrl: URI): PublicKey?
+}
+
+class SdJwtVcVerifier(
+    private val trust: X509CertificateTrust?,
+    private val didResolver: DIDResolver?,
+    private val factory: KtorHttpClientFactory = DefaultHttpClientFactory,
+) {
+    /**
+     * Verifies an SD-JWT (in non enveloped, simple format)
+     * Typically, this is useful to Holder that wants to verify an issued SD-JWT
+     *
+     * @param unverifiedSdJwt the SD-JWT to be verified
+     * @return the verified SD-JWT, if valid. Otherwise, method could raise a [SdJwtVerificationException]
+     * The verified SD-JWT will contain a [JWT][SdJwt.Issuance.jwt] as both string and decoded payload
+     */
+    suspend fun verifyIssuance(
+        unverifiedSdJwt: String,
+    ): Result<SdJwt.Issuance<JwtAndClaims>> =
+        SdJwtVerifier.verifyIssuance(signatureVerifier(), unverifiedSdJwt)
+
+    /**
+     * Verifies an SD-JWT in JWS JSON general of flattened format as defined by RFC7515 and extended by SD-JWT
+     * specification
+     *
+     * Typically, this is useful to Holder that wants to verify an issued SD-JWT
+     *
+     * @param unverifiedSdJwt the SD-JWT to be verified.
+     * A JSON Object that is expected to be in general
+     * or flatten form as defined in RFC7515 and extended by SD-JWT specification.
+     * @return the verified SD-JWT, if valid.
+     * Otherwise, method could raise a [SdJwtVerificationException]
+     * The verified SD-JWT will contain a [JWT][SdJwt.Issuance.jwt] as both string and decoded payload
+     */
+    suspend fun verifyIssuance(
+        unverifiedSdJwt: JsonObject,
+    ): Result<SdJwt.Issuance<JwtAndClaims>> = SdJwtVerifier.verifyIssuance(signatureVerifier(), unverifiedSdJwt)
+
+    private suspend fun signatureVerifier() = sdJwtVcSignatureVerifier(trust, didResolver, factory)
+}
+
+private suspend fun sdJwtVcSignatureVerifier(
+    trust: X509CertificateTrust?,
+    didResolver: DIDResolver?,
+    factory: KtorHttpClientFactory,
+): JwtSignatureVerifier = JwtSignatureVerifier { unverifiedJwt ->
+    try {
+        val signedJwt = SignedJWT.parse(unverifiedJwt)
+        val keySelector = issuerJwsKeySelector(trust, didResolver, signedJwt, factory)
+        checkNotNull(keySelector) { "Failed to resolve issuer public key" }
+        val jwtProcessor = sdJwtVcProcessor(keySelector)
+        jwtProcessor.process(signedJwt, null).asClaims()
+    } catch (e: Throwable) {
+        null
+    }
+}
+
+private suspend fun issuerJwsKeySelector(
+    trust: X509CertificateTrust?,
+    didResolver: DIDResolver?,
+    signedJwt: SignedJWT,
+    factory: KtorHttpClientFactory,
+): JWSKeySelector<SecurityContext>? {
+    val algorithm = requireNotNull(signedJwt.header.algorithm) { "missing 'alg'" }
+    return when (val source = keySource(signedJwt)) {
```

Comment:
@dzarras

Perhaps when() {}   could return a pub key and then assemble a `SingleKeyJWSKeySelector`

Reply:
Perhaps, `issuerJwsKeySelector` could be re-written as `issuerPubKey(source)`

Reply:
Can't do that. In case we have to fetch the Issuer's metadata, the metadata contain a JWKS/JWKS URI which can contain multiple public keys. Hence the need for constructing different JWSKeySelectors.